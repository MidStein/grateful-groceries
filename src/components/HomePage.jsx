import '../styles/HomePage.css';
import { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import groceries from '../groceries.json';

import cart from '../images/cart.svg';
import fruits from '../images/fruits.png';
import vegetables from '../images/vegetables.png';
import meat from '../images/meat.png';
import apple from '../images/apple.png';
import banana from '../images/banana.png';
import cabbage from '../images/cabbage.png';
import chicken from '../images/chicken.png';
import goat from '../images/goat.png';
import lamb from '../images/lamb.png';
import mango from '../images/mango.png';
import potato from '../images/potato.png';
import spinach from '../images/spinach.png';

const categoryImages = {
	fruits,
	vegetables,
	meat
};
const itemImages = {
	apple,
	banana,
	cabbage,
	chicken,
	goat,
	lamb,
	mango,
	potato,
	spinach
};

const HomePage = () => {
	const [items, setItems] = useState({});

	useEffect(() => {
		const qtyDict = {};
		for (let i = 0; i < groceries.categories.length; i += 1) {
			for (let j = 0; j < groceries.categories[i].items.length; j += 1) {
				qtyDict[`${groceries.categories[i].items[j].name}`] = { 
					...groceries.categories[i].items[j],
					qty: 0, 
					isInCart: false, 
				};
			}
		}
		setItems(qtyDict);
	}, []);

	const countCart = () => {
		let cnt = 0;
		for (const item in items) {
			if (items[item].isInCart === true) {
				cnt += items[item].qty;
			}
		}
		return cnt;
	}

	const decrement = (itemName) => {
		if (items[itemName].qty > 0) {
			setItems(() => ({
				...items,
				[itemName]: { ...items[itemName], qty: items[itemName].qty - 1}
			}));
		}
	}
	const increment = (itemName) => {
		setItems(() => ({
			...items,
			[itemName]: { ...items[itemName], qty: items[itemName].qty + 1}
		}));
	}
	const changeHandler = (e, itemName) => {
		setItems(() => ({
			...items,
			[itemName]: { ...items[itemName], qty: Number(e.target.value) }
		}));
	}

	const addToCart = (e, itemName) => {
		e.target.style.backgroundColor = '#22c55e';
		setItems({
			...items,
			[itemName]: { ...items[itemName], isInCart: true }
		})
	}

	return (
		<div className='HomePage'>
			<div 
				style={{ backgroundColor: countCart() > 0 ? '#4ade80' : '#d9f99d' }} 
				className="sticky"
				data-testid="count-cart"
			>
				{countCart()}
				<Link to='/shop' state={{ items: items }}>
					<img src={cart} alt="go to cart" className='cart-button'/>
				</Link>
			</div>
			{groceries.categories.map((category) => (
				<div key={category.category}>
					<div className="grocery-type">
						<img src={categoryImages[`${category.category.toLowerCase()}`]} alt={category.category} />
						<div>{category.category}</div>
					</div>
					<div className='type-group'>
					{category.items.map((item) => (
						<div className="item-card" key={item.name}>
							<img src={itemImages[`${item.name.toLowerCase()}`]} alt={item.name} className="item-pic" />
							<div className="item-name">{item.name}</div>
							<div className="item-price">Rs. {item.price}/-</div>
							<div className="quantity">
								<button onClick={() => decrement(item.name)}>-</button>
								<input type="text" value={items[`${item.name}`]?.qty || 0} onChange={(e) => changeHandler(e, item.name)} />
								<button onClick={() => increment(item.name)}>+</button>
							</div>
							<button onClick={(e) => addToCart(e, item.name)}>Add to <img src={cart} alt='cart'/></button>
						</div>
					))}
					</div>
				</div>
			))}
		</div>
	);
}

export default HomePage;
