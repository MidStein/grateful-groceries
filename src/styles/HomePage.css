.sticky {
	font-size: 30px;
	padding: 6px 0 2px 0;
	position: sticky;
	top: 0;
	right: 0;
	display: flex;
	justify-content: center;
}
.cart-button {
	background-color: #bfdbfe;
	width: 32px;
	height: 32px;
	border: 2px solid #000;
	border-radius: 2px;
	margin-left: 4px;
}
.grocery-type {
	background-color: #d2bab0;
	color: #1e3e8a;
	font-size: 40px;
	height: 48px;
	padding: 4px 0;
	box-shadow: 0 4px 4px rgba(0, 0, 0, 0.25);
	display: flex;
	justify-content: center;
	gap: 4px;
}
.type-group {
	display: flex;
	justify-content: space-around;
	padding: 16px;
}
.item-card {
	background-color: #f2e8e5;
	height: auto;
	border: 2px solid #000;
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	gap: 2px;
	cursor: pointer;
}
.item-card:hover {
	box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.5);
}
.item-pic {
	background-color: #fff;
	width: 250px;
	height: 250px;
	padding: 16px;
}
.item-name {
	font-size: 20px;
	font-weight: 700;
}
.quantity {
	margin-bottom: 2px;
	display: flex;
	gap: 4px;
}
.quantity > input {
	text-align: center;
	width: 64px;
}
.quantity+button {
	background-color: #4ade80;
	border: 1px solid black;
	margin-bottom: 6px;
}
.quantity+button>img {
	width: 16px;
	height: 16px;
	position: relative;
	bottom: -2px;
}
