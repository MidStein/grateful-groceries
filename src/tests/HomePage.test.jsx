import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import HomePage from '../components/HomePage';

describe('Home Page', () => {
	it('renders grocery types and grocery item', () => {
		const { container } = render(
			<BrowserRouter>
				<HomePage />
			</BrowserRouter>
		);

		expect(container).toMatchSnapshot();
	});

	it('renders go to cart clickable image', () => {
		render(
			<BrowserRouter>
				<HomePage />
			</BrowserRouter>
		);

		const cartButton = screen.getByAltText('go to cart');
		expect(cartButton).toBeInTheDocument();
	});

	it('renders price in correct format', () => {
		render(
			<BrowserRouter>
				<HomePage />
			</BrowserRouter>
		);

		const itemPrices = screen.getAllByText(/Rs\. \d+\/-/);
		expect(itemPrices[0]).toBeInTheDocument();
	});

	it('renders add to cart button', () => {
		render(
			<BrowserRouter>
				<HomePage />
			</BrowserRouter>
		);

		const buttons = screen.getAllByRole('button', {name: /Add to /});
		expect(buttons[0]).toBeInTheDocument();
	});
	
	describe('item card input', () => {
		it('value increments/decrements on clicking on respective button', () => {
			render(
				<BrowserRouter>
					<HomePage />
				</BrowserRouter>
			);

			const incBtns = screen.getAllByText('+');
			const decBtns = screen.getAllByText('-');
			const inputs = screen.getAllByRole('textbox');

			userEvent.click(incBtns[0]);
			expect(inputs[0].value).toBe('1');
			userEvent.dblClick(decBtns[0]);
			expect(inputs[0].value).toBe('0');
			userEvent.dblClick(incBtns[0]);
			expect(inputs[0].value).toBe('2');
			userEvent.dblClick(decBtns[0]);
			userEvent.dblClick(decBtns[0]);
			expect(inputs[0].value).toBe('0');
		});

		it('changes according to user typing', () => {
			render(
				<BrowserRouter>
					<HomePage />
				</BrowserRouter>
			);

			const inputs = screen.getAllByRole('textbox');

			userEvent.type(inputs[0], '12');
			expect(inputs[0].value).toBe('12');
			userEvent.clear(inputs[0]);
			userEvent.type(inputs[0], '010');
			expect(inputs[0].value).toBe('10');
		});
	});

	it('renders number of items in cart correctly', () => {
		render(
			<BrowserRouter>
				<HomePage />
			</BrowserRouter>
		);

		const countCart = screen.getByTestId('count-cart');
		const decBtns = screen.getAllByRole('button', { name: '-' });
		const inputs = screen.getAllByRole('textbox');
		const incBtns = screen.getAllByRole('button', { name: '+' });
		const addBtns = screen.getAllByRole('button', { name: /Add to / });

		expect(countCart.textContent).toBe('0');

		userEvent.click(incBtns[0]);
		userEvent.click(addBtns[0]);
		expect(countCart.textContent).toBe('1');

		userEvent.click(addBtns[1]);
		userEvent.dblClick(incBtns[1]);
		expect(countCart.textContent).toBe('3');

		userEvent.click(incBtns[2]);
		userEvent.click(addBtns[2]);
		userEvent.click(decBtns[2]);
		expect(countCart.textContent).toBe('3');

		userEvent.type(inputs[3], '3');
		userEvent.click(addBtns[3]);
		expect(countCart.textContent).toBe('6');
	});
});
