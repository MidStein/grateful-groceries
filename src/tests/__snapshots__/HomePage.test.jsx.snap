// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page renders grocery types and grocery item 1`] = `
<div>
  <div
    class="HomePage"
  >
    <div
      class="sticky"
      data-testid="count-cart"
      style="background-color: rgb(217, 249, 157);"
    >
      0
      <a
        href="/shop"
      >
        <img
          alt="go to cart"
          class="cart-button"
          src="cart.svg"
        />
      </a>
    </div>
    <div>
      <div
        class="grocery-type"
      >
        <img
          alt="Fruits"
          src="fruits.png"
        />
        <div>
          Fruits
        </div>
      </div>
      <div
        class="type-group"
      >
        <div
          class="item-card"
        >
          <img
            alt="Apple"
            class="item-pic"
            src="apple.png"
          />
          <div
            class="item-name"
          >
            Apple
          </div>
          <div
            class="item-price"
          >
            Rs. 
            60
            /-
          </div>
          <div
            class="quantity"
          >
            <button>
              -
            </button>
            <input
              type="text"
              value="0"
            />
            <button>
              +
            </button>
          </div>
          <button>
            Add to 
            <img
              alt="cart"
              src="cart.svg"
            />
          </button>
        </div>
        <div
          class="item-card"
        >
          <img
            alt="Banana"
            class="item-pic"
            src="banana.png"
          />
          <div
            class="item-name"
          >
            Banana
          </div>
          <div
            class="item-price"
          >
            Rs. 
            40
            /-
          </div>
          <div
            class="quantity"
          >
            <button>
              -
            </button>
            <input
              type="text"
              value="0"
            />
            <button>
              +
            </button>
          </div>
          <button>
            Add to 
            <img
              alt="cart"
              src="cart.svg"
            />
          </button>
        </div>
        <div
          class="item-card"
        >
          <img
            alt="Mango"
            class="item-pic"
            src="mango.png"
          />
          <div
            class="item-name"
          >
            Mango
          </div>
          <div
            class="item-price"
          >
            Rs. 
            70
            /-
          </div>
          <div
            class="quantity"
          >
            <button>
              -
            </button>
            <input
              type="text"
              value="0"
            />
            <button>
              +
            </button>
          </div>
          <button>
            Add to 
            <img
              alt="cart"
              src="cart.svg"
            />
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="grocery-type"
      >
        <img
          alt="Vegetables"
          src="vegetables.png"
        />
        <div>
          Vegetables
        </div>
      </div>
      <div
        class="type-group"
      >
        <div
          class="item-card"
        >
          <img
            alt="Spinach"
            class="item-pic"
            src="spinach.png"
          />
          <div
            class="item-name"
          >
            Spinach
          </div>
          <div
            class="item-price"
          >
            Rs. 
            10
            /-
          </div>
          <div
            class="quantity"
          >
            <button>
              -
            </button>
            <input
              type="text"
              value="0"
            />
            <button>
              +
            </button>
          </div>
          <button>
            Add to 
            <img
              alt="cart"
              src="cart.svg"
            />
          </button>
        </div>
        <div
          class="item-card"
        >
          <img
            alt="Cabbage"
            class="item-pic"
            src="cabbage.png"
          />
          <div
            class="item-name"
          >
            Cabbage
          </div>
          <div
            class="item-price"
          >
            Rs. 
            80
            /-
          </div>
          <div
            class="quantity"
          >
            <button>
              -
            </button>
            <input
              type="text"
              value="0"
            />
            <button>
              +
            </button>
          </div>
          <button>
            Add to 
            <img
              alt="cart"
              src="cart.svg"
            />
          </button>
        </div>
        <div
          class="item-card"
        >
          <img
            alt="Potato"
            class="item-pic"
            src="potato.png"
          />
          <div
            class="item-name"
          >
            Potato
          </div>
          <div
            class="item-price"
          >
            Rs. 
            80
            /-
          </div>
          <div
            class="quantity"
          >
            <button>
              -
            </button>
            <input
              type="text"
              value="0"
            />
            <button>
              +
            </button>
          </div>
          <button>
            Add to 
            <img
              alt="cart"
              src="cart.svg"
            />
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="grocery-type"
      >
        <img
          alt="Meat"
          src="meat.png"
        />
        <div>
          Meat
        </div>
      </div>
      <div
        class="type-group"
      >
        <div
          class="item-card"
        >
          <img
            alt="Chicken"
            class="item-pic"
            src="chicken.png"
          />
          <div
            class="item-name"
          >
            Chicken
          </div>
          <div
            class="item-price"
          >
            Rs. 
            60
            /-
          </div>
          <div
            class="quantity"
          >
            <button>
              -
            </button>
            <input
              type="text"
              value="0"
            />
            <button>
              +
            </button>
          </div>
          <button>
            Add to 
            <img
              alt="cart"
              src="cart.svg"
            />
          </button>
        </div>
        <div
          class="item-card"
        >
          <img
            alt="Lamb"
            class="item-pic"
            src="lamb.png"
          />
          <div
            class="item-name"
          >
            Lamb
          </div>
          <div
            class="item-price"
          >
            Rs. 
            90
            /-
          </div>
          <div
            class="quantity"
          >
            <button>
              -
            </button>
            <input
              type="text"
              value="0"
            />
            <button>
              +
            </button>
          </div>
          <button>
            Add to 
            <img
              alt="cart"
              src="cart.svg"
            />
          </button>
        </div>
        <div
          class="item-card"
        >
          <img
            alt="Goat"
            class="item-pic"
            src="goat.png"
          />
          <div
            class="item-name"
          >
            Goat
          </div>
          <div
            class="item-price"
          >
            Rs. 
            150
            /-
          </div>
          <div
            class="quantity"
          >
            <button>
              -
            </button>
            <input
              type="text"
              value="0"
            />
            <button>
              +
            </button>
          </div>
          <button>
            Add to 
            <img
              alt="cart"
              src="cart.svg"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
